File: go/sequence.go
package sequence

// Direction is an alias for order pattern.
type Direction string

const (
	// Increasing order
	Increasing = Direction("increasing")
	// Decreasing order
	Decreasing = Direction("decreasing")
	// Stable order
	Stable = Direction("stable")
)

// Analyze detects the direction of sequence.
func Analyze(input []int) Direction {
	if len(input) <= 1 {
		return Stable
	}

	seen := make(map[Direction]struct{})
	last := input[0]

	for i := 1; i < len(input); i++ {
		if input[i] == last {
			return Stable
		}

		if input[i] > last {
			seen[Increasing] = struct{}{}
		} else {
			seen[Decreasing] = struct{}{}
		}

		last = input[i]
	}

	_, iok := seen[Increasing]
	_, dok := seen[Decreasing]

	if iok && dok {
		return Stable
	}

	if iok {
		return Increasing
	}

	return Decreasing
}

File: ruby/sequence.rb
require 'set'

def analyze(input=[])
  return :stable if input.length < 2

  seen = Set.new
  last = input[0]

  (1..input.length-1).each do |i|
    return :stable if input[i] == last

    if input[i] > last
      seen.add(:increasing)
    else
      seen.add(:decreasing)
    end

    last = input[i]
  end

  return :stable if seen.include?(:increasing) && seen.include?(:decreasing)

  return :increasing if seen.include?(:increasing)

  :decreasing
end

def print(name, input)
  puts "name:\t#{name}"
  puts "input:\t#{input.join(' ')}"
  puts "output:\t#{analyze(input)}"
  puts '=' * 80
end

print('Increasing', [1, 2, 3, 4, 5])
print('Decreasing', [5, 4, 3, 2, 1, 0, -1, -2, -3])
print('Stable', [1, 2, 3, 2, 1, 4, 5])
print('Empty', [])
print('Single', [1])
print('EqualWhenIncreasing', [1, 2, 2])
print('EqualWhenDecreasing', [5, 4, 4])
